gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java

```diff
@@ -0,0 +1,327 @@
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import org.hl7.fhir.dstu3.model.CodeableConcept;
+import org.hl7.fhir.dstu3.model.Coding;
+import org.hl7.fhir.dstu3.model.DomainResource;
+import org.hl7.fhir.dstu3.model.Encounter;
+import org.hl7.fhir.dstu3.model.Encounter.EncounterLocationComponent;
+import org.hl7.fhir.dstu3.model.Encounter.EncounterParticipantComponent;
+import org.hl7.fhir.dstu3.model.Encounter.EncounterStatus;
+import org.hl7.fhir.dstu3.model.Identifier;
+import org.hl7.fhir.dstu3.model.ListResource;
+import org.hl7.fhir.dstu3.model.ListResource.ListEntryComponent;
+import org.hl7.fhir.dstu3.model.Meta;
+import org.hl7.fhir.dstu3.model.Patient;
+import org.hl7.fhir.dstu3.model.Period;
+import org.hl7.fhir.dstu3.model.Reference;
+import org.hl7.v3.CD;
+import org.hl7.v3.CsNullFlavor;
+import org.hl7.v3.IVLTS;
+import org.hl7.v3.IVXBTS;
+import org.hl7.v3.RCMRMT030101UK04Author;
+import org.hl7.v3.RCMRMT030101UK04Component;
+import org.hl7.v3.RCMRMT030101UK04Component02;
+import org.hl7.v3.RCMRMT030101UK04Component3;
+import org.hl7.v3.RCMRMT030101UK04Component4;
+import org.hl7.v3.RCMRMT030101UK04CompoundStatement;
+import org.hl7.v3.RCMRMT030101UK04EhrComposition;
+import org.hl7.v3.RCMRMT030101UK04EhrExtract;
+import org.hl7.v3.RCMRMT030101UK04EhrFolder;
+import org.hl7.v3.RCMRMT030101UK04LinkSet;
+import org.hl7.v3.RCMRMT030101UK04Participant2;
+import org.hl7.v3.TS;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.util.CollectionUtils;
+
+import lombok.RequiredArgsConstructor;
+import uk.nhs.adaptors.pss.translator.util.DateFormatUtil;
+import uk.nhs.adaptors.pss.translator.util.EhrResourceExtractorUtil;
+
+@Service
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class EncounterMapper {
+    private static final List<String> INVALID_CODES = List.of("196401000000100", "196391000000103");
+    private static final String ENCOUNTER_META_PROFILE = "https://fhir.nhs.uk/STU3/StructureDefinition/CareConnect-GPC-Encounter-1";
+    private static final String PRACTITIONER_REFERENCE_PREFIX = "Practitioner/";
+    private static final String LOCATION_REFERENCE_PREFIX = "Location/";
+    private static final String LOCATION_REFERENCE_SUFFIX = "-LOC";
+    private static final String PERFORMER_SYSTEM = "http://hl7.org/fhir/v3/ParticipationType";
+    private static final String PERFORMER_CODE = "PPRF";
+    private static final String PERFORMER_DISPLAY = "primary performer";
+    private static final String RECORDER_SYSTEM = "https://fhir.nhs.uk/STU3/CodeSystem/GPConnect-ParticipantType-1";
+    private static final String RECORDER_CODE = "REC";
+    private static final String RECORDER_DISPLAY = "recorder";
+    private static final String TOPIC_CLASS_CODE = "TOPIC";
+    private static final String CATEGORY_CLASS_CODE = "CATEGORY";
+    private static final String ENCOUNTER_KEY = "encounters";
```

Comment:
Perhaps we could use ResourceType enum as key?

Reply:
As per our conversation, this isnt possible due to Consultaiton, Topic and Categories all being List Resources.